File: 2022/day3.rb
#!/usr/bin/env ruby

class Day3
  def self.run(input_file)
    runner = Day3.new(input_file)
    puts "Part 1: #{runner.part_one}"
    puts "Part 2: #{runner.part_two}"
  end

  def initialize(filename)
    @input = File.readlines(filename, chomp: true)
  end

  def part_one
    overlap = []
    @input.each do |rucksack|
      half = (rucksack.length.to_f / 2).ceil
      one = rucksack[0, half]
      two = rucksack[half, rucksack.length]

      overlap << (one.chars & two.chars)
    end

    overlap.flatten.map { |c| score_value(c) }.sum
  end

  def part_two
    overlap = []
    @input.each_slice(3) do |slice|
      overlap << (slice[0].chars & slice[1].chars & slice[2].chars)
    end
    overlap.flatten.map { |c| score_value(c) }.sum
  end

  def score_value(char)
    if char =~ /[A-Z]/
      char.ord - 38
    else
      char.ord - 96
    end
  end
end

if ARGV.empty?
  Day3.run 'input/day3.txt'
else
  Day3.run 'test/day3.txt'
end

File: 2022/day3.go
package main

import (
	"bufio"
	"os"
	"strings"
)

func main() {
	var input *os.File
	var err error

	if len(os.Args[1:]) > 0 {
		input, err = os.Open("test/day3.txt")
	} else {
		input, err = os.Open("input/day3.txt")
	}

	if err != nil {
		os.Exit(1)
	}

	scoreValue("A")
	scoreValue("Z")
	scoreValue("a")
	scoreValue("z")
	println("Part 1: ", partOne(input))
	input.Seek(0, 0)
	println("Part 1: ", partTwo(input))
}

func partOne(file *os.File) int {
	score := 0
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	for scanner.Scan() {
		rucksack := scanner.Text()
		packages := strings.SplitAfter(rucksack, "")
		middle := len(packages) / 2
		one := packages[0:middle]
		two := packages[middle:]

		for _, charOne := range one {
			for _, charTwo := range two {
				if charOne == charTwo {
					println(charOne, scoreValue(charOne))
					score += scoreValue(charOne)
				}
			}
		}
	}

	return score
}

func partTwo(file *os.File) int {
	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	for scanner.Scan() {
	}

	return 0
}

func scoreValue(char string) int {
	value := int([]rune(char)[0])

	if char == strings.ToUpper(char) {
		return value - 38
	} else {
		return value - 96
	}
}
